### PR TITLE
[Bun.sql] Support TLS

### PR DESCRIPTION
**Pull request by Jarred-Sumner**
### What does this PR do?

The following works:
```ts
import { test, expect } from "bun:test";
import { getSecret } from "harness";
import { sql as SQL } from "bun";

const TLS_POSTGRES_DATABASE_URL = getSecret("TLS_POSTGRES_DATABASE_URL");

test("tls (explicit)", async () => {
  const sql = new SQL({
    url: TLS_POSTGRES_DATABASE_URL!,
    tls: true,
    adapter: "postgresql",
  });

  const [{ one, two }] = await sql`SELECT 1 as one, '2' as two`;
  expect(one).toBe(1);
  expect(two).toBe("2");
});

test("tls (implicit)", async () => {
  const [{ one, two }] = await SQL`SELECT 1 as one, '2' as two`;
  expect(one).toBe(1);
  expect(two).toBe("2");
});
```

It successfully connected to Neon and with TLS.

### How did you verify your code works?

There is a test

### CODE REVIEW

**Comment by robobun** [comment]
@Jarred-Sumner, your commit 04295f90123192b9fa58c8f4433a82936e92f9d5 is building: [`#6313`](https://buildkite.com/bun/bun/builds/6313)

File: packages/bun-usockets/src/crypto/openssl.c [hunk 1]
```diff
@@ -231,6 +231,10 @@ struct us_internal_ssl_socket_t *ssl_on_open(struct us_internal_ssl_socket_t *s,
   if (is_client) {
     SSL_set_renegotiate_mode(s->ssl, ssl_renegotiate_explicit);
     SSL_set_connect_state(s->ssl);
+
+    if (sni) {
+      SSL_set_tlsext_host_name(s->ssl, sni);
```
Comment (cirospaciari) [inline packages/bun-usockets/src/crypto/openssl.c:236]: normally we do this inside on_open (that happens before the on_handshake) but I think is a OK change, should we do something on server side?

File: src/js/bun/sql.ts [hunk 2]
```diff
@@ -507,6 +569,10 @@ function SQL(o) {
 
 var lazyDefaultSQL;
 var defaultSQLObject = function sql(strings, ...values) {
+  if (new.target) {
```
Comment (cirospaciari) [inline src/js/bun/sql.ts:572]: TIL https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/new.target

File: src/sql/postgres.zig [hunk 3]
```diff
@@ -1059,48 +1134,97 @@ pub const PostgresSQLConnection = struct {
...
-        event_loop.enter();
-        defer event_loop.exit();
-        this.flushData();
+        this.start();
+    }
+
+    pub fn onHandshake(this: *PostgresSQLConnection, success: i32, ssl_error: uws.us_bun_verify_error_t) void {
+        debug("onHandshake: {d} {d}", .{ success, ssl_error.error_no });
+        if (success != 1) {
+            if (this.tls_config.reject_unauthorized == 1) {
```
Comment (cirospaciari) [inline src/sql/postgres.zig]: should we use `BoringSSL.checkServerIdentity` to match fetch and other clients behavior? (checking altname and IP)

Comment (cirospaciari) [reply src/sql/postgres.zig]: fixed in https://github.com/oven-sh/bun/pull/15217/commits/2fa15c9492854b8fb2b00415f2732bfe15df7f14

**Review by cirospaciari — APPROVED** [review]

**Pull request merged**